### PR TITLE
feat(compile): Add `--no-terminal` to compile command

### LINKED ISSUE #17990: Feature request: Option to prevent terminal from opening after compile (on windows)
Windows applications differ from Linux and MacOS in that they can either be a console or GUI application.
Currently Deno is a console application by default, which carries over when using `deno compile`

However there are use cases for which you would not want a terminal window to open, see [this discussion](https://github.com/denoland/deno/discussions/11638). In the comments a user provides a script to change the relevant entry in the PE header to make any application a GUI application.

Another use case being possible webview support, for which it would not be a good user experience to see a terminal pop up.
[See latest comments in this discussion](https://github.com/denoland/deno/discussions/3234?sort=new#discussioncomment-5149598)

So I propose to add a flag, like `--no-terminal`, to `deno compile` which changes the header so it won't open the terminal.

### PR DESCRIPTION
**Pull request by JasperVanEsveld**
This adds the feature requested in: closes #17990 

When compiling checks if `--no-terminal` is set and the target is windows.
If so it changes `mSubsystem` of the PE header to 2 in the original binary.
Which indicates to windows that it is a GUI application, check [here](https://wiki.osdev.org/PE) for more info on the PE header.

### CODE REVIEW

**Comment by justinmchase** [comment]
I feel like there should be a solution for this for the other platforms as well. Do you know what would be involved with that?

**Comment by JasperVanEsveld** [comment]
Do you mean for other OS's?
Because this is a windows only problem, Linux and MacOS won't force a terminal to open.
Maybe the flag would need to be named to reflect that, any suggestions?

**Comment by justinmchase** [comment]
Well I think that in mac you have to build a .app file. Even though its not adding a flag to the process the compile step has to do something special. Something different but similar to linux, if you are using the `--compile` flag on deno to compile a stand alone app it will make a terminal app by default. Even if it doesn't do the same thing technically per OS it seems like there should be a flag which, instead of a simple executable would actually produce an OS dependent "windowed app".

**Comment by justinmchase** [comment]
Not that you have to solve this all in this PR but for example I'd think that this same flag you might produce a .app file instead for mac, something like this:

https://apple.stackexchange.com/a/407885

**Comment by JasperVanEsveld** [comment]
That does sound related yes.

For MacOS the bare minimum seems fairly simple, just uses a specific dir structure.
However I don't have access to a Mac, so I won't be able to test if it actually works.

The other points from your link (icon and launchpad) could be addressed in a future PR.

Do you know what needs to change for Linux? I wasn't able to find anything...

File: cli/tools/standalone.rs [hunk 1]
```diff
@@ -209,6 +223,18 @@ async fn create_standalone_binary(
   Ok(final_bin)
 }
 
+fn set_windows_binary_to_gui(mut bin: Vec<u8>) -> Vec<u8> {
```
Comment (justinmchase) [inline cli/tools/standalone.rs:226]: I feel like this could use some code comments including a documentation link. I think I can tell what you're doing, flipping a few bits which are probably flags on the header but I gotta say it does look pretty magical, it would be nice to know what is happening precisely.

For example whate are the numbers 60-64 and 92? It may even be nice to use the windows flag constants for these values.

**Review by aapoalas — CHANGES_REQUESTED** [review]
Left few comments. In addition to those, a flags test is also necessary for merging. Doesn't need to be much, just a test similar to `compile_with_flags` exercising `no_terminal: true` path as well. If additions to the `Args` build are made to check for Windows-only usage of the flag then tests for those interactions would be good as well.

Also, PR needs a rebase apparently.

**Comment by JasperVanEsveld** [comment]
Thanks for the feedback, will look into it in the coming week when I have time

File: cli/args/flags.rs [hunk 2]
```diff
@@ -923,6 +924,11 @@ fn compile_subcommand<'a>() -> Command<'a> {
           "aarch64-apple-darwin",
         ]),
     )
+    .arg(
+      Arg::new("no-terminal")
+        .long("no-terminal")
+        .help("Prevents terminal from opening up on windows"),
```
[the remark this answers is not in the document]
Comment (JasperVanEsveld) [reply cli/args/flags.rs]: Yes, the help message is a bit wordy.
Limiting the flag to just windows is not really an option, as you can set the compile target with `--target`.
So on linux/mac you can compile to windows, which could use the `--no-terminal` flag.

What do you think of `Hide terminal on Windows`?
That is what I changed it to for now, does not have the repeating window though it is less descriptive.

File: cli/tools/standalone.rs [hunk 3]
```diff
@@ -209,6 +223,18 @@ async fn create_standalone_binary(
...
+  let header_start = LittleEndian::read_i32(&bin[60..64]);
```
[the remark this answers is not in the document]
Comment (JasperVanEsveld) [reply cli/tools/standalone.rs]: Thanks, I didn't know `align_to` was a thing.
I removed the crate and used your suggestion.
Also wrapped it into two functions, `read_as` and `read_as_mut`.
As I need to read multiple times and want to prevent size mismatches.

**Comment by JasperVanEsveld** [comment]
Sorry that it took a bit longer, didn't really have time last week

File: cli/tools/standalone.rs [hunk 4]
```diff
@@ -223,16 +222,74 @@ async fn create_standalone_binary(
...
+  let slice = &mut bin[range];
+  let (prefix, value, suffix) = unsafe { slice.align_to_mut::<T>() };
+  if prefix.len() != 0 || suffix.len() != 0 || value.len() == 0 {
+    bail!("Could not align bytes to a slice of Ts")
+  }
+  Ok(value)
+}
+
+/// This function sets the subsystem field in the PE header to 2 (GUI subsystem)
+/// For more information about the PE header: https://learn.microsoft.com/en-us/windows/win32/debug/pe-format
```
Comment (justinmchase) [inline cli/tools/standalone.rs:236]: 👍

**Review by justinmchase — APPROVED** [review]

**Comment by JasperVanEsveld** [comment]
@aapoalas 
Could you take a look at this again?
Whenever you have time of course

**Review by aapoalas — APPROVED** [review]
Looks good to me, just requires rebase.

Sorry it took so long to get back to you, and thank you very much for the contribution!

**Comment by JasperVanEsveld** [comment]
> Looks good to me, just requires rebase.
> 
> Sorry it took so long to get back to you, and thank you very much for the contribution!

No worries, should be up to date again!

**Review by aapoalas — APPROVED** [review]

**Review by justinmchase — APPROVED** [review]

**Comment by JasperVanEsveld** [comment]
It has been a few days since the changes were approved
Does anything else need to happen or does it just take some time before PRs are merged?

**Comment by JasperVanEsveld** [comment]
@aapoalas 
Maybe you are busy, but I would love if this got merged

Also, merged the main branch because it was out of date, now the tests need to be run again
Should I just not do that as long as there are no conflicts?

**Review by aapoalas — APPROVED** [review]
Looks good to me, @bartlomieju do you want to bless this (conditional that tests pass but they should be fine).

**Comment by aapoalas** [comment]
Alright, sorry for taking so long with this but now we have it merged! Thank you for the contribution!

**Comment by lino-levan** [comment]
Exciting PR merge for sure.
